src/core/lib/surface/call.c

```diff
@@ -1384,8 +1397,8 @@ static void receiving_initial_metadata_ready(grpc_exec_ctx *exec_ctx,
     }
   }
 
-  call->has_initial_md_been_received = true;
-  if (call->saved_receiving_stream_ready_bctlp != NULL) {
+  if (!gpr_atm_acq_cas(&call->recv_state, RECV_NONE,
```

Comment:
Unfortunately, all our [`gpr_atm_xxx_cas()`](https://github.com/grpc/grpc/blob/master/include/grpc/impl/codegen/atm_gcc_atomic.h#L64) APIs  use `__ATOMIC_RELAXED` memory order in case of failure.    i.e they all are designed for cases like:
```
if (gpr_atm_xxx_cas(...)) {
     // Do something
 }
```

but not for the opposite. i.e
```
if ( ! gpr_atm_xxx_cas(...)) {
     // Do something
 }
```
I think this is the case for adding new variants to our gpr_atm_xxx_cas APIs (where we can specify __ATOMIC_ACQUIRE mem order for failure case)

cc/ @dklempner @ctiller @vjpai  to add any comments here.